timemachine/fe/chiral_utils.py

```diff
@@ -484,3 +489,338 @@ def make_chiral_flip_heatmaps(simulation_result, atom_map):
     assert mol_a_chiral_conflicts.shape == (len(simulation_result.frames), len(simulation_result.frames[0]))
 
     return mol_a_chiral_conflicts, mol_b_chiral_conflicts
+
+
+def get_num_connected_components(num_atoms: int, bonds: Collection[Tuple[int, int]]) -> int:
+    g = nx.Graph()
+    g.add_nodes_from(range(num_atoms))
+    g.add_edges_from(bonds)
+    return len(list(nx.connected_components(g)))
+
+
+def canonicalize_chiral_atom_idxs(idxs):
+    i, j, k, l = idxs
+    rotations = [(j, k, l), (l, j, k), (k, l, j)]
+    jj, kk, ll = min(rotations)
+    return i, jj, kk, ll
+
+
+def setup_end_state_harmonic_bond_and_chiral_potentials(
```

Comment:
Maybe add a TODO comment here to decompose this into harmonic bond + chiral restraint setup and move the harmonic bond part back to `single_topology`?